### PR TITLE
Qmaps 667

### PR DESCRIPTION
**Pull request by xem**

### CODE REVIEW

File: src/adapters/scene.js [hunk 1]
```diff
@@ -143,6 +147,92 @@ Scene.prototype.initMapBox = function () {
...
+  // sw
+  if(this.isPointInBounds(bbox._sw, viewport)){
+    bboxIsNear = true;
+  }
+
+  return bboxIsNear
+
+}
+
+Scene.prototype.fitBbox = function(bbox, padding){
```
Comment (Julien-laville) [inline src/adapters/scene.js]: can we merge this function with fitMap, and let the choice to the refactored function ? @amatissart

File: src/adapters/scene.js [hunk 2]
```diff
@@ -143,6 +147,92 @@ Scene.prototype.initMapBox = function () {
...
+
+  padding = padding || {left: 0, top: 0, right: 0, bottom: 0};
+
+  console.log(padding);
```
Comment (Julien-laville) [inline src/adapters/scene.js]: lint to remove

File: src/adapters/scene.js [hunk 3]
```diff
@@ -143,6 +147,92 @@ Scene.prototype.initMapBox = function () {
...
+
+  // Animate if the zoom is big enough and if the BBox is (partially or fully) in the extended viewport
+  var animate = false;
+
+  if(this.mb.getZoom() > 10 && this.isBBoxInExtendedViewport(bbox)){
```
Comment (Julien-laville) [inline src/adapters/scene.js]: this logic already exist

File: src/adapters/direction_api.js [hunk 4]
```diff
@@ -8,7 +8,7 @@ const geometries = 'geojson'
 export default class DirectionApi {
   static async search(start, end, vehicle, exclude = '') {
     const directionsUrl = `https://api.mapbox.com/directions/v5/mapbox/${vehicle}/${poiToMapBoxCoordinates(start)}%3B${poiToMapBoxCoordinates(end)}.json`
-    return await Ajax.get(directionsUrl, {language : getLang().locale, exclude : exclude, geometries : geometries, steps : true, access_token : token, alternatives : true, overview : OVERVIEW_SETING})
+    return await Ajax.get(directionsUrl, {language : getLang().locale, exclude : exclude, geometries : geometries, steps : true, access_token : token, alternatives : true, annotations: "congestion", overview : OVERVIEW_SETING})
```
[the remark this answers is not in the document]
Comment (xem) [reply src/adapters/direction_api.js]: K, but I'm disabling it as long as we don't know how to display traffic properly

File: src/panel/direction/road_map_panel.js [hunk 5]
```diff
@@ -82,4 +83,16 @@ export default class RoadMapPanel {
...
+  mouseOverStep(i){
+    fire("highlight_step", i);
+  }
+
+  mouseOutStep(i){
+    fire("unhighlight_step", i);
+  }
+
+  clickStep(step){
+    fire("zoom_step", step, {top: 20, right: 20, bottom: 40, left: 450 });
```
[the remark this answers is not in the document]
Comment (xem) [reply src/panel/direction/road_map_panel.js]: yes but left is not enough, even on desktop, when we zoom on a step there should be padding on all sides. I agree this object should be factorized somewhere though

File: src/adapters/scene_direction.js [hunk 6]
```diff
@@ -30,46 +35,91 @@ export default class SceneDirection {
...
+
+    listen('unhighlight_step', (step) => {
+      this.unhighlight_step(step);
     })
   }
 
+  showMarkerSteps(){
+
+    // Hide previously drawn steps markers
+    if(this.markersSteps.length > 0){
```
[the remark this answers is not in the document]
Comment (xem) [reply src/adapters/scene_direction.js]: for-in too actually. fixed

File: src/adapters/scene_direction.js [hunk 7]
```diff
@@ -30,46 +35,91 @@ export default class SceneDirection {
...
+
+    this.markersSteps = []
+
+    for (var step = 1; step < this.steps.length - 1; step++) {
+
+      const markerStep = document.createElement('div')
+      markerStep.className = 'itinerary_marker_step'
+      markerStep.onclick = (function(step){
+          return function() {
+              fire("zoom_step", step)
```
[the remark this answers is not in the document]
Comment (xem) [reply src/adapters/scene_direction.js]: I think not, we're in a closure, the goal is to fire the event when the marker is clicked. The marker can't call this.computeBBox, can it?

File: src/adapters/scene.js [hunk 8]
```diff
@@ -143,6 +147,92 @@ Scene.prototype.initMapBox = function () {
...
+
+  // Get viewport bounds
+  var viewport = this.mb.getBounds();
+
+  // Compute "width", "height"
+  var width = viewport._ne.lng - viewport._sw.lng;
+  var height = viewport._ne.lat - viewport._sw.lat;
+
+  // Compute extended viewport
+  viewport._ne.lng += width;
```
[the remark this answers is not in the document]
Comment (xem) [reply src/adapters/scene.js]: I'd rather not let every function call choose its viewport scale. No?

File: src/adapters/scene.js (hunk 3 above)
Comment (xem) [reply src/adapters/scene.js]: yep, it's here until we merge the functions

File: src/adapters/scene_direction.js [hunk 9]
```diff
@@ -30,46 +35,91 @@ export default class SceneDirection {
...
+      }
+    }
+
+    this.markersSteps = []
+
+    for (var step = 1; step < this.steps.length - 1; step++) {
+
+      const markerStep = document.createElement('div')
+      markerStep.className = 'itinerary_marker_step'
+      markerStep.onclick = (function(step){
```
[the remark this answers is not in the document]
Comment (xem) [reply src/adapters/scene_direction.js]: I can replace "function" with "=>", but is there a way to simplify more than that?

File: src/adapters/scene_direction.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (Julien-laville) [reply src/adapters/scene_direction.js]: this is overwritten by the function keyword (sort of) but arrow function keep the correct 'this' context

File: src/adapters/scene.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (Julien-laville) [reply src/adapters/scene.js]: the idea is not to let this as public, but integrate this function in fitMap, fitMap can set the correct ratio maybe with some explicit const

File: src/adapters/scene.js (hunk 1 above)
Comment (amatissart) [reply src/adapters/scene.js]: yes, it should be possible to refactor these conditions on zoom and viewport.  
Why not simply call this `fitBbox` from `fitMap` too ?

File: src/libs/actions.js [hunk 10]
```diff
@@ -22,6 +22,50 @@ class Click{
   }
 }
 
+class MouseOver{
```
[the remark this answers is not in the document]
Comment (Julien-laville) [reply src/libs/actions.js]: `import Telemetry from "./telemetry";


class Action {
  constructor(id, action, eventName) {
    this.action = action
    this.id = id
    this.eventName = eventName
  }

  toString() {
    return ` on${this.eventName}="call4Action(event, ${this.id})" `
  }

  exec() {
    if(this.telemetry) {
      this.telemetry.add()
    }
    this.action.method.call(this.action.ctx, this.action.args)
  }

  addTelemetry(message) {
    this.telemetry = new Telemetry(message)
    return this
  }
}

/**
 * bind html native listener to panel action
 */

(function actions() {
  const actions = new Map()
  const supportedActions = ['mouseover', 'click', 'mouseout']
  /**
   *
   * @param method call back function
   * @param ctx "this"
   * @returns {string}
   */

  supportedActions.forEach((actionName) => {
    window[actionName] = function (method, ctx, options = {}) {
      const action = {
        id : actions.size,
        method : method,
        ctx : ctx,
        args : options
      }
      let clickAction = new Action(action.id, action, actionName)
      actions.set(action.id, clickAction)
      return clickAction
    }
  })

  window.call4Action = function (event, id) {
    event.stopPropagation()
    const action = actions.get(id)
    action.exec()
  }
})()
`

File: src/adapters/scene.js [hunk 11]
```diff
@@ -143,30 +147,99 @@ Scene.prototype.initMapBox = function () {
   })
 }
 
-Scene.prototype.fitMap = function(poi, options = {}) {
```
Comment (Julien-laville) [inline src/adapters/scene.js]: fitMap & fitBBox seems to be still not merged

File: src/adapters/scene.js [hunk 12]
```diff
@@ -143,30 +147,99 @@ Scene.prototype.initMapBox = function () {
...
+  viewport._ne.lng += width;
+  viewport._ne.lat += height;
+
+  viewport._sw.lng -= width;
+  viewport._sw.lat -= height;
+
+  // Check bounds:
+
+  // Lng between -180 and 180 (wraps: 180 + 1 = -179)
+  if(viewport._ne.lng < 180) viewport._ne.lng += 360;
```
Comment (Julien-laville) [inline src/adapters/scene.js]: so, no helper ?
